File: tools/binding_compliance/conformance/families/papyrus_monitor.py
"""Papyrus counters require full read, tail-from-end, idle and reset semantics."""

from collections.abc import Mapping
from functools import partial
from typing import Any

from ..coverage import CoveragePredicate, FamilyCoveragePolicy


def _stats(value: object) -> bool:
    """Require complete nonnegative counters and the native ratio relationship."""
    if not isinstance(value, Mapping) or set(value) != {
        "dumps",
        "stacks",
        "warnings",
        "errors",
        "lines",
        "ratio",
    }:
        return False
    if not all(
        type(value[key]) is int and value[key] >= 0
        for key in ("dumps", "stacks", "warnings", "errors", "lines")
    ):
        return False
    ratio = (
        value["dumps"] / value["stacks"] if value["dumps"] and value["stacks"] else 0.0
    )
    return value["ratio"] == f"{ratio:.3f}"


def _observed(exists: bool, observation: Mapping[str, Any]) -> bool:
    """Reject missing stages, replayed idle data and failure disguised as an empty log."""
    if (
        set(observation)
        != {
            "exists",
            "error",
            "initial",
            "tailStart",
            "updated",
            "idle",
            "afterReset",
            "finalContent",
        }
        or observation["exists"] is not exists
    ):
        return False
    stages = ("initial", "tailStart", "updated", "idle", "afterReset")
    if not exists:
        return (
            observation["error"] == "missing"
            and observation["finalContent"] is None
            and all(observation[key] is None for key in stages)
        )
    if (
        observation["error"] is not None
        or not all(_stats(observation[key]) for key in stages)
        or not isinstance(observation["finalContent"], str)
    ):
        return False
    return (
        all(
            observation["tailStart"][key] == 0
            for key in ("dumps", "stacks", "warnings", "errors", "lines")
        )
        and observation["idle"] == observation["updated"]
        and all(
            observation["afterReset"][key]
            == observation["initial"][key] + observation["updated"][key]
            for key in ("dumps", "stacks", "warnings", "errors", "lines")
        )
        and observation["afterReset"]["lines"]
        == len(observation["finalContent"].splitlines())
    )


def _full(observation: Mapping[str, Any]) -> bool:
    """Require complete Node-exported full-file counters or an actual missing-file error."""
    if set(observation) != {"stats", "error", "content"}:
        return False
    if observation["error"] == "missing":
        return observation["stats"] is None and observation["content"] is None
    value = observation["stats"]
    return (
        observation["error"] is None
        and isinstance(observation["content"], str)
        and isinstance(value, Mapping)
        and set(value) == {"dumps", "stacks", "warnings", "errors", "lines"}
        and all(type(count) is int and count >= 0 for count in value.values())
        and value["lines"] == len(observation["content"].splitlines())
    )


PAPYRUS_MONITOR_COVERAGE_POLICY = FamilyCoveragePolicy(
    "papyrus-monitor",
    tuple(
        CoveragePredicate(
            "monitor-stages" if exists else "missing-log",
            "papyrus-monitor.observe",
            "papyrus-monitor.observe",
            "monitor-stages",
            ("PapyrusAnalyzer", "PapyrusStats") if exists else ("PapyrusAnalyzer",),
            partial(_observed, exists),
            runtime_operations=(
                None,
                "__init__",
                "analyze_full",
                "analyze_to_string",
                "log_exists",
                "log_path",
                "stats",
                "reset",
                "start_monitoring",
                "check_for_updates",
                "dumps_to_stacks_ratio",
                "papyrus_analyzer_new",
                "papyrus_analyze_full",
                "papyrus_log_exists",
                "papyrus_reset",
                "papyrus_start_monitoring",
                "papyrus_check_updates",
                "papyrus_logging",
            )
            if exists
            else (
                "analyze_full",
                "papyrus_analyze_full",
                "log_exists",
                "papyrus_log_exists",
            ),
        )
        for exists in (True, False)
    )
    + (
        CoveragePredicate(
            "full-file",
            "papyrus-monitor.full",
            "papyrus-monitor.full",
            "full-stats",
            ("PapyrusStats",),
            _full,
            binding_obligation_ids=("parity:node:scanlog-papyrus-analysis",),
            runtime_operations=("analyzePapyrusLog",),
        ),
    ),
)

File: business-logic/classic-scanlog-core/tests/semantic_conformance/papyrus_monitor.rs
//! Native Papyrus full-analysis and tail-monitor lifetime observations.

use super::{RunnerResult, invalid, text};
use classic_scanlog_core::papyrus::{PapyrusAnalyzer, PapyrusError, PapyrusStats};
use serde_json::{Value, json};
use std::{fs, io::Write};

/// Projects portable counters and the public ratio query, excluding clock timestamps.
fn stats(value: &PapyrusStats) -> Value {
    json!({"dumps":value.dumps,"stacks":value.stacks,"warnings":value.warnings,
        "errors":value.errors,"lines":value.lines_processed,"ratio":format!("{:.3}",value.dumps_to_stacks_ratio())})
}

/// Runs full, tail, append, idle and reset stages in an owned disposable directory.
pub(super) fn observe(fixture: &Value) -> RunnerResult<Value> {
    let temporary = tempfile::tempdir()?;
    let path = temporary.path().join("Papyrus.0.log");
    if !fixture["content"].is_null() {
        fs::write(&path, text(&fixture["content"])?.as_bytes())?;
    }
    let mut analyzer = PapyrusAnalyzer::new(path.clone());
    if fixture["operation"] == "full" {
        return match analyzer.analyze_full() {
            Ok(value) => Ok(
                json!({"stats":{"dumps":value.dumps,"stacks":value.stacks,"warnings":value.warnings,"errors":value.errors,"lines":value.lines_processed},"error":null,"content":fs::read_to_string(path)?}),
            ),
            Err(PapyrusError::LogNotFound(_)) => {
                Ok(json!({"stats":null,"error":"missing","content":null}))
            }
            Err(error) => Err(error.into()),
        };
    }
    let mut result = json!({"exists":analyzer.log_exists(),"error":null,"initial":null,
        "tailStart":null,"updated":null,"idle":null,"afterReset":null,"finalContent":null});
    if analyzer.log_path() != path || stats(analyzer.stats()) != stats(&PapyrusStats::new()) {
        return Err(invalid("new Papyrus analyzer has wrong initial state").into());
    }
    result["initial"] = match analyzer.analyze_full() {
        Ok(value) => stats(&value),
        Err(PapyrusError::LogNotFound(_)) => {
            result["error"] = json!("missing");
            return Ok(result);
        }
        Err(error) => return Err(error.into()),
    };
    let value = analyzer.stats();
    let expected = format!(
        "NUMBER OF DUMPS    : {}\nNUMBER OF STACKS   : {}\nDUMPS/STACKS RATIO : {:.3}\nNUMBER OF WARNINGS : {}\nNUMBER OF ERRORS   : {}\nLINES PROCESSED    : {}",
        value.dumps,
        value.stacks,
        value.dumps_to_stacks_ratio(),
        value.warnings,
        value.errors,
        value.lines_processed
    );
    if analyzer.analyze_to_string() != expected {
        return Err(invalid("Papyrus summary disagrees with statistics").into());
    }
    analyzer.start_monitoring()?;
    if analyzer.check_for_updates()?.is_some() {
        return Err(invalid("tail start replayed old lines").into());
    }
    result["tailStart"] = stats(analyzer.stats());
    let appended = text(&fixture["append"])?;
    fs::OpenOptions::new()
        .append(true)
        .open(&path)?
        .write_all(appended.as_bytes())?;
    if let Some((lines, _)) = analyzer.check_for_updates()?
        && lines != appended.lines().map(str::to_owned).collect::<Vec<_>>()
    {
        return Err(invalid("tail update changed lines").into());
    }
    result["updated"] = stats(analyzer.stats());
    if analyzer.check_for_updates()?.is_some() {
        return Err(invalid("idle poll replayed lines").into());
    }
    result["idle"] = stats(analyzer.stats());
    analyzer.reset();
    result["afterReset"] = stats(&analyzer.analyze_full()?);
    result["finalContent"] = json!(fs::read_to_string(path)?);
    Ok(result)
}
